File: source-app/flutter_app/lib/features/settings/users/user_last_active.dart
import 'package:intl/intl.dart';

/// Last-active copy for user list + profile (no "No recent activity").
abstract final class UserLastActive {
  static String label(String? lastActiveIso, {String? createdAtIso}) {
    final d = _parseUtc(lastActiveIso);
    if (d == null) {
      final created = _parseUtc(createdAtIso);
      if (created != null &&
          DateTime.now().difference(created.toLocal()) <
              const Duration(days: 14)) {
        return 'Created recently';
      }
      return 'Never active';
    }
    final local = d.toLocal();
    final now = DateTime.now();
    final today = DateTime(now.year, now.month, now.day);
    final day = DateTime(local.year, local.month, local.day);
    if (day == today) {
      return 'Today ${DateFormat.jm().format(local)}';
    }
    if (day == today.subtract(const Duration(days: 1))) {
      return 'Yesterday ${DateFormat.jm().format(local)}';
    }
    if (now.difference(local) < const Duration(days: 7)) {
      return DateFormat('EEE h:mm a').format(local);
    }
    return DateFormat.yMMMd().add_jm().format(local);
  }

  static bool isOnlineNow(String? lastActiveIso) {
    final d = _parseUtc(lastActiveIso);
    if (d == null) return false;
    return DateTime.now().toUtc().difference(d) < const Duration(minutes: 5);
  }

  static DateTime? _parseUtc(String? iso) {
    if (iso == null || iso.isEmpty) return null;
    return DateTime.tryParse(iso)?.toUtc();
  }
}

abstract final class UserRoleStyle {
  static String displayRole(String? role) {
    final r = (role ?? '').toLowerCase();
    return switch (r) {
      'owner' => 'Owner',
      'admin' => 'Admin',
      'manager' => 'Manager',
      'staff' => 'Staff',
      _ => r.isEmpty ? '—' : r[0].toUpperCase() + r.substring(1),
    };
  }

  static String statusLabel({required bool blocked, required bool active}) {
    if (blocked) return 'Blocked';
    if (active) return 'Active';
    return 'Inactive';
  }
}

File: new-app/frontend/src/features/users/userLastActive.ts
/**
 * Last-active + role/status labels — user_last_active.dart UserLastActive / UserRoleStyle.
 */

function parseUtc(iso: string | null | undefined): Date | null {
  if (iso == null || iso === "") return null;
  const d = new Date(iso);
  return Number.isNaN(d.getTime()) ? null : d;
}

function formatJm(d: Date): string {
  return d.toLocaleTimeString(undefined, {
    hour: "numeric",
    minute: "2-digit",
  });
}

/** UserLastActive.label */
export function userLastActiveLabel(
  lastActiveIso: string | null | undefined,
  createdAtIso?: string | null,
): string {
  const d = parseUtc(lastActiveIso);
  if (d == null) {
    const created = parseUtc(createdAtIso);
    if (created != null) {
      const ageMs = Date.now() - created.getTime();
      if (ageMs < 14 * 24 * 60 * 60 * 1000) {
        return "Created recently";
      }
    }
    return "Never active";
  }
  const local = d;
  const now = new Date();
  const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
  const day = new Date(local.getFullYear(), local.getMonth(), local.getDate());
  if (day.getTime() === today.getTime()) {
    return `Today ${formatJm(local)}`;
  }
  const yesterday = new Date(today);
  yesterday.setDate(yesterday.getDate() - 1);
  if (day.getTime() === yesterday.getTime()) {
    return `Yesterday ${formatJm(local)}`;
  }
  if (now.getTime() - local.getTime() < 7 * 24 * 60 * 60 * 1000) {
    return local.toLocaleString(undefined, {
      weekday: "short",
      hour: "numeric",
      minute: "2-digit",
    });
  }
  return local.toLocaleString(undefined, {
    year: "numeric",
    month: "short",
    day: "numeric",
    hour: "numeric",
    minute: "2-digit",
  });
}

/** UserLastActive.isOnlineNow — 5 minute window */
export function userIsOnlineNow(lastActiveIso: string | null | undefined): boolean {
  const d = parseUtc(lastActiveIso);
  if (d == null) return false;
  return Date.now() - d.getTime() < 5 * 60 * 1000;
}

/** UserRoleStyle.displayRole */
export function displayUserRole(role: string | null | undefined): string {
  const r = (role ?? "").toLowerCase();
  switch (r) {
    case "owner":
      return "Owner";
    case "admin":
      return "Admin";
    case "manager":
      return "Manager";
    case "staff":
      return "Staff";
    default:
      if (!r) return "—";
      return r[0]!.toUpperCase() + r.slice(1);
  }
}

/** UserRoleStyle.statusLabel */
export function userStatusLabel(opts: {
  blocked: boolean;
  active: boolean;
}): string {
  if (opts.blocked) return "Blocked";
  if (opts.active) return "Active";
  return "Inactive";
}
